### PR TITLE
commitlog: recalculate disk footprint on delete_segment exceptions

### LINKED ISSUE #9348: Exceptions in commitlog::segment_manager::delete_segments could cause footprint counters to loose track
If we get errors/exceptions in `delete_segments` we can (and probably will) loose track of disk footprint counters. This can in turn, if using hard limits, cause us to block indefinitely on segment allocation since we might think we have larger footprint than we actually do. 

Of course, if we actually fail deleting a segment, it is 100% true that we still technically hold this disk footprint (now unreachable), but for cases where for example outside forces (or wacky tests) delete a file behind our backs, this might not be true. One could also argue that our footprint is the segments and file names we keep track of, and the rest is exterior sludge. 

In any case, if we have any exceptions in `delete_segments`, we should recalculate disk footprint based on current state, and restart all `new_segment` paths etc.

### PR DESCRIPTION
**Pull request by elcallio**
If we get errors/exceptions in delete_segments we can (and probably will) loose track of disk footprint counters. This can in turn, if using hard limits, cause us to block indefinitely on segment allocation since we might think we have larger footprint than we actually do.

Of course, if we actually fail deleting a segment, it is 100% true that we still technically hold this disk footprint (now unreachable), but for cases where for example outside forces (or wacky tests) delete a file behind our backs, this might not be true. One could also argue that our footprint is the segments and file names we keep track of, and the rest is exterior sludge.

In any case, if we have any exceptions in delete_segments, we should recalculate disk footprint based on current state, and restart all new_segment paths etc.

Fixes #9348

(Note: this is based on previous PR #9344 - so shows these commits as well. Actual changes are only the latter two).

### CODE REVIEW

**Comment by avikivity** [comment]
I think the previous series causes dtest errors and may need to be dequeued. Let me check.

**Comment by elcallio** [comment]
Updated to sync harder on recalculation, based on dtest shenanigans.

**Comment by elcallio** [comment]
ping?

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1525,8 +1525,11 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
 
         f = make_checked_file(commit_error_handler, std::move(f));
     } catch (...) {
-        ep = std::current_exception();
-        commit_error_handler(ep);
+        try {
+            commit_error_handler(std::current_exception());
+        } catch (...) {
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:1539]: This throws only in extremely non-default configurations, yes?

Not saying it shouldn't be corrected for those cases, just trying to understand the scope of the problem.

**Comment by elcallio** [comment]
> 
> 
> Won't exceptions in delete_segments stop the node due to I/O errors?

delete_segments just logs any exception, and might also abort the recycled list iff we failed a recycle rename (to wake up waiters). It neither propagates exceptions not aborts anything. Technically, iff we fail deletes, it is pretty bad and maybe an abort condition, but so far we have not treated it as such. Perhaps because spurious delete fails _can_ happen due to user messing with file system at the wrong time, but the node can potentially go on after it (albeit with a dangled file). Not sure on best approach...

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:1539]: This patch is really part of different PR (see top comment), but to answer the question: commit_error_handler rethrows internally, and wraps + rethrows iff system_error, or aborts node iff configured as such. We need to always propagate the error out of catch to do async cleanup. 
It should be very rare to reach this, yes.

**Pull request closed without merging**